libs/client/src/streaming.ts

```diff
@@ -0,0 +1,251 @@
+
+const EVENT_STREAM_TIMEOUT = 15 * 1000;
+
+type FalStreamEventType = 'message' | 'error' | 'done';
+
+type EventHandler = (event: any) => void;
+
+/**
+ * The class representing a streaming response. With t
+ */
+class FalStream<Input, Output> {
+  // properties
+  url: string;
+  options: StreamOptions<Input>;
+
+  // support for event listeners
+  private listeners: Map<FalStreamEventType, EventHandler[]> = new Map();
+  private buffer: Output[] = [];
+
+  // local state
+  private currentData: Output | undefined = undefined;
+  private lastEventTimestamp = 0;
+  private streamClosed = false;
+  private donePromise: Promise<Output>;
+
+  constructor(url: string, options: StreamOptions<Input>) {
+    this.url = url;
+    this.options = options;
+    this.donePromise = new Promise<Output>((resolve, reject) => {
+      if (this.streamClosed) {
+        reject(
+          new ApiError({
+            message: 'Streaming connection is already closed.',
+            status: 400,
+            body: undefined,
+          })
+        );
+      }
+      this.on('done', (data) => {
+        this.streamClosed = true;
+        resolve(data);
+      });
+      this.on('error', (error) => {
+        this.streamClosed = true;
+        reject(error);
+      });
+    });
+    this.start();
+  }
+
+  private start = async () => {
+    try {
+      const response = await fetch(this.url, {
+        method: 'POST',
+        headers: {
+          accept: 'text/event-stream',
+          'content-type': 'application/json',
+        },
+        body: JSON.stringify(this.options.input),
+      });
+      this.handleResponse(response);
+    } catch (error) {
+      this.handleError(error);
+    }
+  };
+
+  private handleResponse = async (response: Response) => {
+    if (!response.ok) {
+      try {
+        // we know the response failed, call the response handler
+        // so the exception gets converted to ApiError correctly
+        await defaultResponseHandler(response);
+      } catch (error) {
+        this.emit('error', error);
+      }
+      return;
+    }
+
+    const body = response.body;
+    if (!body) {
+      this.emit(
+        'error',
+        new ApiError({
+          message: 'Response body is empty.',
+          status: 400,
+          body: undefined,
+        })
+      );
+      return;
+    }
+    const decoder = new TextDecoder('utf-8');
+    const reader = response.body.getReader();
+
+    const parser = createParser((event) => {
+      if (event.type === 'event') {
+        const data = event.data;
+
+        try {
+          const parsedData = JSON.parse(data);
+          this.buffer.push(parsedData);
+          this.currentData = parsedData;
+          this.emit('message', parsedData);
+        } catch (e) {
+          this.emit('error', e);
+        }
+      }
+    });
+
+    const timeout = this.options.timeout ?? EVENT_STREAM_TIMEOUT;
+
+    const readPartialResponse = async () => {
+      const { value, done } = await reader.read();
```

Comment:
Where does `done` come from?

Reply:
From the data reader itself, since it's reading a stream, it knows when the response is closed and returns the `done` flag.